### ISSUE TITLE
extract should fail on extra fields

### ISSUE THREAD

**Issue by charpov** [json4s/json4s#650]
`extract` succeeds when a JSON object has *more* fields than the targeted `case` class.  Is this a desirable feature?  Is there an option to fail unless there is an exact match?  It seems dangerous to ignore extraneous JSON fields silently like that.

**Comment by charpov** [comment]
My current workaround:

```scala
  def safeExtract[A](json: JValue)(implicit formats: Formats, mf: scala.reflect.Manifest[A]): A = {
    val x = json.extract[A]
    if (json == decompose(x)) x else throw new MappingException("not exact")
  }
```

**Comment by magnolia-k** [comment]
@charpov 

The current situation is the intended behavior. If you need to provide arguments for all existing fields, you will have less freedom in handling JSON.

However, you can also understand that there are situations where you want to make sure that they match, depending on your use case. In that case, it should be another method like `extractStrict`.
Patches are welcome.

**Comment by charpov** [comment]
I'm writing tests for a server and I'd like to make sure the replies are exactly what they're supposed to be.  This is what I'm using for now:
```scala
  def safeExtract[A](json: JValue)(implicit formats: Formats, mf: scala.reflect.Manifest[A]): A = {
    val x = json.extract[A]
    if (json == decompose(x)) x
    else {
      val Diff(_, _, extra) = json.diff(decompose(x))
      throw new MappingException(s"extra JSON: `${compactRender(extra)}`")
    }
  }
```

**Comment by magnolia-k** [comment]
Certainly, there are situations where you want to check that there are no extra fields in test applications.

We cannot guarantee that they can be added immediately, but we will keep the issue open.

**Comment by mccartney** [comment]
Looks like a duplicate of #432
